Automattic/jetpack
# Decoupling register and authentication actions in the main connection flow

This PR decouples the registration and authorization steps in the main connection flow, in a way we can choose whether to offer in-place authorization after the site is register and we have some information from the server.

NOTE: This PR does not cover secondary users authentication, reconnection or in-line user authorization from the dashboard when the site is in site connection mode. All those will still work as usual and will be handled in following PRs.

Basically what this PR does is to move the decision on whether to use the in-place flow or not one step further, after the user registration.

Before, we made this decision at the very start of the flow and all the registration process would take place in Calypso if in-place connection was disabled for any reason. Now, the site registration is always done using our local REST endpoint and only once the site is registered we choose if the authorization will be in-line or if the user will be redirected to Calypso.

Also, we add the possibility for the WPCOM servers to inform an alternate authorization URL, that would take the user to a different flow.

![Captura de Tela 2021-05-13 às 11 03 53](https://user-images.githubusercontent.com/971483/118136848-f0418400-b3da-11eb-852d-6c7379724999.png)


#### Changes proposed in this Pull Request:

* Changes the main connection flow and chooses between in-place or calypso flows only after the site registration

#### Jetpack product discussion


p9dueE-2R7-p2

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Sandbox your site and apply D61359-code to your sandbox
* Test the main connection flow. The in-place connection should work as expected
* Test the main connection flow in safari or other browser that does not support 3rd pary cookies. It should work as always
* Modify the code in your sandbox to not allow in-place authorization
* Test the main connection flow again. In all cases, you should be redirected to Calypso. Test the flow until the end and make sure you are redirected back to the dashboard after choosing "Start for free"

##### Testing alternate auth URL

Run the tests below with 3 different values for the alternate_authentication_url:

* a valid redirect source (you can choose one from Mission Control `/jetpack-crew/redirects/`
* a URL under the wordpress.com domain (e.g. https://wordpress.com/support)
* a URL outside our domains - in this case you should be redirected to the jetpack.com homepage)

* Modify the code on your sandbox again and inform an alternate authorization URL
* Test the main connection flow again and make sure you are redirected to the alternate URL once the registration is complete
* Go back to your dashboard and make sure you are connected at a site level
